File: Python3/452. Minimum Number of Arrows to Burst Balloons.py
class Solution:
    def findMinArrowShots(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        # points.sort(key=lambda x: x[1])
        # ans, end = 0, -float('inf')
        # for point in points:
        #     if point[0] > end:
        #         ans += 1
        #         end = point[1]
        # return ans
        
        # 实现discuss中的此问题模板做法
        points.sort(key=lambda x: x[0])
        ans, end = 0, -float('inf')
        for point in points:
            if point[0] > end:
                ans += 1
                end = point[1]
            else:
                end = min(end, point[1])
        return ans

File: Golang/452. Minimum Number of Arrows to Burst Balloons.go
// func findMinArrowShots(points [][]int) int {
//     sort.Slice(points, func(i, j int) bool {
//         if points[i][1] == points[j][1] {
//             return points[i][0] > points[j][0]
//         }
//         return points[i][1] < points[j][1]
//     })

//     res := 0
//     end := -1 << 32
    
//     for _, point := range points {
//         if point[0] > end {
//             res++
//             end = point[1]
//         }
//     }
    
//     return res
// }

func findMinArrowShots(points [][]int) int {
    sort.Slice(points, func(i, j int) bool {
        // if points[i][0] == points[j][0] {
        //     return points[i][1] > points[j][1]
        // }
        return points[i][0] < points[j][0]
    })

    res := 0
    end := -1 << 32
    
    for _, point := range points {
        if point[0] > end {
            res++
            end = point[1]
        } else {
            if end > point[1] {
                end = point[1]
            }
        }
    }
    
    return res
}